RFCs/0009-Fn::ForEach.md

```diff
@@ -0,0 +1,1461 @@
+        RoutingRules:
+          - RoutingRuleCondition:
+              HttpErrorCodeReturnedEquals: '404'
+              KeyPrefixEquals: out1/
+            RedirectRule:
+              HostName: ec2-11-22-333-44.compute-1.amazonaws.com
+              ReplaceKeyPrefixWith: report-404/
+    DeletionPolicy: Retain
+    UpdateReplacePolicy: Retain
+  S3BucketC:
+    Type: 'AWS::S3::Bucket'
+    Properties:
+      AccessControl: PublicRead
+      MetricsConfigurations:
+        - Id: EntireBucketC
+      WebsiteConfiguration:
+        IndexDocument: index.html
+        ErrorDocument: error.html
+        RoutingRules:
+          - RoutingRuleCondition:
+              HttpErrorCodeReturnedEquals: '404'
+              KeyPrefixEquals: out1/
+            RedirectRule:
+              HostName: ec2-11-22-333-44.compute-1.amazonaws.com
+              ReplaceKeyPrefixWith: report-404/
+    DeletionPolicy: Retain
+    UpdateReplacePolicy: Retain
+Outputs:
+  S3BucketAArn:
+    Value: !GetAtt [S3BucketA, Arn]
+  S3BucketADomainName:
+    Value: !GetAtt [S3BucketA, DomainName]
+  S3BucketAWebsiteURL:
+    Value: !GetAtt [S3BucketA, WebsiteURL]
+  S3BucketBArn:
+    Value: !GetAtt [S3BucketB, Arn]
+  S3BucketBDomainName:
+    Value: !GetAtt [S3BucketB, DomainName]
+  S3BucketBWebsiteURL:
+    Value: !GetAtt [S3BucketB, WebsiteURL]
+  S3BucketCArn:
+    Value: !GetAtt [S3BucketC, Arn]
+  S3BucketCDomainName:
+    Value: !GetAtt [S3BucketC, DomainName]
+  S3BucketCWebsiteURL:
+    Value: !GetAtt [S3BucketC, WebsiteURL]
+```
+
+#### Use case 6: Reference a specific resource among replicated resources
+
+In this example, customer is referencing to a resource created in a Fn::ForEach loop using the respective generated Logical IDs.
+
+#### JSON
+
+```json
+{
+    "AWSTemplateFormatVersion": "2010-09-09",
+    "Transform": "AWS::LanguageExtensions",
+    "Mappings": {
+        "Instances": {
+            "InstanceType": {
+                "B": "m5.4xlarge",
+                "C": "c5.2xlarge"
+            },
+            "ImageId": {
+                "A": "ami-id1"
+            }
+        }
+    },
+    "Resources": {
+        "Fn::ForEach::Instances": [
+            "Identifier",
+            ["A", "B", "C"],
+            {
+                "Instance${Identifier}": {
+                    "Type": "AWS::EC2::Instance",
+                    "Properties": {
+                        "InstanceType": {
+                            "Fn::FindInMap": ["Instances", "InstanceType", {"Ref": "Identifier"}, {"DefaultValue": "m5.xlarge"}]
+                        },
+                        "ImageId": {
+                            "Fn::FindInMap": ["Instances", "ImageId", {"Ref": "Identifier"}, {"DefaultValue": "ami-id-default"}]
+                        }
+                    }
+                }
+            }
+        ]
+    },
+    "Outputs": {
+        "SecondInstanceId": {
+            "Description": "Instance Id for InstanceB",
+            "Value": {"Ref": "InstanceB"}
```

Comment:
This feels gross, having to reference a generated identifier, relative to something that is more like `!Ref Fn::ForEach::Instances[1]`

Reply:
Yeah the ForEach name has to be a unique identifier in the template namespace (see Constraint 3 in RFC), where based on the number of asks for the feature, !Ref on the ForEach name, which would return a list of logicalId’s generated by ForEach, can be implemented in the future (see GitHub issue [here](https://github.com/aws-cloudformation/cfn-language-discussion/issues/118) with details on Additional Enhancements to ForEach)